File: src/main/wallpaper.ts
/**
 * Custom wallpaper lifecycle — MAIN only (PA-8: fs/net/dialogs live here).
 *
 * The user picks a background image (local file via the native dialog, an
 * explicit local path, or an http(s) URL). MAIN copies/downloads the image into
 * the MANAGED folder `{userData}/wallpaper/` and persists ONLY the managed file
 * name in the settings store (`settings.wallpaperFileName`) — the user's
 * original path / the remote URL are never referenced again, so the wallpaper
 * keeps working if the source moves or goes offline (mirrors how UWP apps copy
 * picked assets into ApplicationData.Current.LocalFolder).
 *
 * Lifecycle rules:
 *   - REPLACE: the new image is written FIRST, the setting flips to the new
 *     name, then the previous managed file is deleted — a crash mid-replace can
 *     orphan at most one file, never break the active wallpaper.
 *   - CLEAR: the setting is emptied, then the managed file is deleted. No
 *     orphan accumulation: there is at most ONE managed wallpaper file.
 *   - Persistence + cross-window propagation ride the EXISTING settings path:
 *     setSettings() broadcasts EvtSettingsChanged to every window, and each
 *     renderer re-fetches the image via `wallpaper.get()` when the file name in
 *     its settings bag changes (no new push channel needed).
 *
 * Security (URL download):
 *   - http(s) only; the response content-type MUST be an allowed raster image
 *     type (SVG is deliberately excluded — it can embed script; raster formats
 *     are inert), and the byte stream is capped at MAX_WALLPAPER_BYTES while
 *     reading (a missing/lying Content-Length cannot blow past the cap).
 *   - Remote content is only ever STORED and later served to the renderer as a
 *     `data:` image URL (the page CSP already allows `img-src data:`) — it is
 *     never executed, never loaded as a document.
 *
 * Pure helpers (name/extension/mime validation) are exported for vitest; the
 * electron-touching functions (dialog/net/settings) stay e2e-territory, same
 * split as window-bounds.ts / mru.ts.
 */

import { copyFile, mkdir, readFile, rm, stat, writeFile } from 'node:fs/promises';
import { extname, join } from 'node:path';
import { app, dialog, net, BrowserWindow } from 'electron';
import type { Result, WallpaperState } from '../shared/ipc-contract.js';
import { getSettings, setSettings } from './settings.js';

/** Managed subfolder under userData holding the (single) active wallpaper. */
const WALLPAPER_DIR_NAME = 'wallpaper';

/** Hard size cap for any wallpaper source (local stat / download stream). */
export const MAX_WALLPAPER_BYTES = 20 * 1024 * 1024; // 20 MB

/**
 * Allowed raster image extensions ↔ MIME types. SVG is EXCLUDED on purpose
 * (scriptable format; see module header). This single table drives extension
 * validation (local picks), content-type validation (downloads), and the
 * data-URL mime when serving the file back to the renderer.
 */
const IMAGE_TYPES: readonly { ext: string; mime: string }[] = [
  { ext: 'png', mime: 'image/png' },
  { ext: 'jpg', mime: 'image/jpeg' },
  { ext: 'jpeg', mime: 'image/jpeg' },
  { ext: 'webp', mime: 'image/webp' },
  { ext: 'gif', mime: 'image/gif' },
  { ext: 'bmp', mime: 'image/bmp' },
  { ext: 'avif', mime: 'image/avif' }
];

function errMsg(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}

/**
 * Resolve the userData root. Honors the e2e override (`NOTEPADS_E2E_USERDATA`)
 * BEFORE touching electron's `app`, exactly like settings.ts / mru.ts — this is
 * also what lets the fs-lifecycle unit tests run electron-free.
 */
function userDataRoot(): string {
  const override = process.env['NOTEPADS_E2E_USERDATA'];
  if (override && override.length > 0) return override;
  return app.getPath('userData');
}

/** Absolute path of the managed wallpaper folder. */
function wallpaperDir(): string {
  return join(userDataRoot(), WALLPAPER_DIR_NAME);
}

// ---------------------------------------------------------------------------
//  Pure helpers (exported for vitest)
// ---------------------------------------------------------------------------

/** Lower-cased extension (no dot) of a path/name, '' when absent. */
export function imageExtensionOf(pathOrName: string): string {
  return extname(pathOrName).replace(/^\./, '').toLowerCase();
}

/** True when `ext` (no dot, any case) is an allowed raster image extension. */
export function isAllowedImageExtension(ext: string): boolean {
  const lower = ext.toLowerCase();
  return IMAGE_TYPES.some((t) => t.ext === lower);
}

/**
 * Map a response Content-Type (possibly with `; charset=` suffix) to the
 * canonical managed extension, or null when it is not an allowed image type.
 * This is the download-path gate: a URL serving text/html (or anything
 * non-image) is REJECTED regardless of its file extension.
 */
export function extensionForContentType(contentType: string | null): string | null {
  if (!contentType) return null;
  const mime = contentType.split(';')[0].trim().toLowerCase();
  const found = IMAGE_TYPES.find((t) => t.mime === mime);
  return found ? found.ext : null;
}

/** MIME type for a managed extension (the data-URL header when serving). */
export function mimeForExtension(ext: string): string | null {
  const found = IMAGE_TYPES.find((t) => t.ext === ext.toLowerCase());
  return found ? found.mime : null;
}

/**
 * Compose a fresh managed file name. The timestamp keeps successive wallpapers
 * distinct (replace never overwrites in place, and a changed name doubles as a
 * cache-buster for the renderer), `nowMs` is injectable for deterministic tests.
 */
export function buildWallpaperFileName(ext: string, nowMs: number = Date.now()): string {
  return `wallpaper-${nowMs}.${ext.toLowerCase()}`;
}

/**
 * Validate a PERSISTED wallpaper file name before resolving it under the
 * managed folder. Settings.json is user-editable on disk, so the name must
 * match exactly what buildWallpaperFileName produces — this forbids path
 * separators / `..` traversal and any extension outside the allowed set.
 */
export function isSafeWallpaperFileName(name: string): boolean {
  const m = /^wallpaper-\d+\.([a-z0-9]+)$/.exec(name);
  return m !== null && isAllowedImageExtension(m[1]);
}

// ---------------------------------------------------------------------------
//  Managed-folder fs primitives (electron-free; unit-tested with a temp dir)
// ---------------------------------------------------------------------------

/**
 * Write `bytes` as a new managed wallpaper file inside `dir` (created on
 * demand). Returns the absolute path written. Name must already be validated.
 */
export async function writeManagedWallpaper(
  dir: string,
  fileName: string,
  bytes: Buffer
): Promise<string> {
  await mkdir(dir, { recursive: true });
  const target = join(dir, fileName);
  await writeFile(target, bytes);
  return target;
}

/**
 * Delete a previously managed wallpaper file, best-effort. Refuses unsafe
 * names (defense in depth — only names this module generated are removable)
 * and never throws: replace/clear must not fail because the old file is
 * already gone or briefly locked.
 */
export async function deleteManagedWallpaper(dir: string, fileName: string): Promise<void> {
  if (!fileName || !isSafeWallpaperFileName(fileName)) return;
  await rm(join(dir, fileName), { force: true }).catch(() => {});
}

// ---------------------------------------------------------------------------
//  IPC-facing API (electron-touching)
// ---------------------------------------------------------------------------

/**
 * Mutation serializer — guards the "at most ONE managed file" invariant.
 *
 * Without it, two interleaved activations (a URL set racing a Browse pick:
 * each window's IPC handler awaits independently) can BOTH read the same
 * `previous` name; the loser's freshly written file then ends up neither
 * active nor deleted — a permanent orphan. Chaining every read-modify-delete
 * section through one module-level promise makes the previous-name read and
 * the old-file delete atomic with respect to other mutations. Only the FAST
 * critical section (settings write + unlink) is serialized — slow downloads/
 * copies happen before entering the queue, so a slow URL fetch never blocks a
 * concurrent local pick. Exported for the vitest ordering test.
 */
let wallpaperMutationChain: Promise<unknown> = Promise.resolve();
export function enqueueWallpaperMutation<T>(op: () => Promise<T>): Promise<T> {
  // Run after whatever is queued, whether it settled ok or not (a failed
  // mutation must never wedge the chain — second arg runs on rejection too).
  const run = wallpaperMutationChain.then(op, op);
  // Swallow for the CHAIN only; callers still observe `run`'s real outcome.
  wallpaperMutationChain = run.catch(() => {});
  return run;
}

/** Read the current persisted wallpaper file name ('' when none). */
async function currentFileName(): Promise<string> {
  const r = await getSettings();
  return r.ok ? r.data.wallpaperFileName : '';
}

/**
 * Activate `fileName`: persist it (broadcasts EvtSettingsChanged to every
 * window via the settings path), then delete the previous managed file. Write
 * order (new file already on disk → setting → delete old) means a crash can
 * orphan at most one stale file, never dangle the active setting. The whole
 * read→persist→delete section runs inside the mutation queue so concurrent
 * activations can never both read the same `previous` (orphan race).
 */
function activate(fileName: string): Promise<Result<WallpaperState>> {
  return enqueueWallpaperMutation(async () => {
    const previous = await currentFileName();
    const persisted = await setSettings({ wallpaperFileName: fileName });
    if (!persisted.ok) return persisted;
    if (previous && previous !== fileName) {
      await deleteManagedWallpaper(wallpaperDir(), previous);
    }
    return getWallpaper();
  });
}

/**
 * Current wallpaper as a data URL. A missing/manually-deleted file resolves to
 * the empty state (reads never fail the app — same policy as settings reads).
 */
export async function getWallpaper(): Promise<Result<WallpaperState>> {
  try {
    const fileName = await currentFileName();
    if (!fileName || !isSafeWallpaperFileName(fileName)) {
      return { ok: true, data: { fileName: '', dataUrl: null } };
    }
    const mime = mimeForExtension(imageExtensionOf(fileName));
    if (!mime) return { ok: true, data: { fileName: '', dataUrl: null } };
    const bytes = await readFile(join(wallpaperDir(), fileName));
    return {
      ok: true,
      data: { fileName, dataUrl: `data:${mime};base64,${bytes.toString('base64')}` }
    };
  } catch {
    // File vanished underneath the setting (manual cleanup) — treat as unset.
    return { ok: true, data: { fileName: '', dataUrl: null } };
  }
}

/**
 * Copy a local image into the managed folder and activate it. Validates the
 * extension against the allowed raster set and enforces the size cap BEFORE
 * copying. The user's original file is left untouched and never referenced.
 */
export async function setWallpaperFromPath(path: string): Promise<Result<WallpaperState>> {
  try {
    const ext = imageExtensionOf(path);
    if (!isAllowedImageExtension(ext)) {
      return { ok: false, error: `Unsupported image type: .${ext || '(none)'}` };
    }
    const info = await stat(path);
    if (!info.isFile()) return { ok: false, error: 'Not a file' };
    if (info.size > MAX_WALLPAPER_BYTES) {
      return { ok: false, error: `Image exceeds the ${MAX_WALLPAPER_BYTES / (1024 * 1024)}MB limit` };
    }
    const fileName = buildWallpaperFileName(ext);
    const dir = wallpaperDir();
    await mkdir(dir, { recursive: true });
    await copyFile(path, join(dir, fileName));
    return activate(fileName);
  } catch (e) {
    return { ok: false, error: errMsg(e) };
  }
}

/**
 * Download an http(s) image into the managed folder and activate it. Uses
 * Electron's `net.fetch` (Chromium network stack — proxy/cert aware). The
 * content-type gate and the streaming byte cap both run BEFORE anything is
 * written to disk; remote bytes are stored as an inert image file only.
 */
export async function setWallpaperFromUrl(url: string): Promise<Result<WallpaperState>> {
  try {
    let parsed: URL;
    try {
      parsed = new URL(url);
    } catch {
      return { ok: false, error: 'Invalid URL' };
    }
    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
      return { ok: false, error: 'Only http(s) URLs are supported' };
    }

    const res = await net.fetch(parsed.toString(), { redirect: 'follow' });
    if (!res.ok) return { ok: false, error: `Download failed (HTTP ${res.status})` };

    const ext = extensionForContentType(res.headers.get('content-type'));
    if (!ext) {
      return { ok: false, error: 'URL did not return a supported image type' };
    }
    const declaredLength = Number(res.headers.get('content-length') ?? 0);
    if (declaredLength > MAX_WALLPAPER_BYTES) {
      return { ok: false, error: `Image exceeds the ${MAX_WALLPAPER_BYTES / (1024 * 1024)}MB limit` };
    }

    // Stream with a hard cap: a missing/lying Content-Length cannot overrun.
    if (!res.body) return { ok: false, error: 'Empty response body' };
    const reader = res.body.getReader();
    const chunks: Buffer[] = [];
    let total = 0;
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      total += value.byteLength;
      if (total > MAX_WALLPAPER_BYTES) {
        await reader.cancel().catch(() => {});
        return { ok: false, error: `Image exceeds the ${MAX_WALLPAPER_BYTES / (1024 * 1024)}MB limit` };
      }
      chunks.push(Buffer.from(value));
    }
    if (total === 0) return { ok: false, error: 'Empty response body' };

    const fileName = buildWallpaperFileName(ext);
    await writeManagedWallpaper(wallpaperDir(), fileName, Buffer.concat(chunks));
    return activate(fileName);
  } catch (e) {
    return { ok: false, error: errMsg(e) };
  }
}

/**
 * Prompt for a local image via MAIN's native open dialog, then run the same
 * copy/activate path. Cancel resolves `null` as a normal success (file.openDialog
 * convention). PA-8: the dialog never reaches the renderer.
 */
export async function pickWallpaper(): Promise<Result<WallpaperState | null>> {
  try {
    const focused = BrowserWindow.getFocusedWindow() ?? BrowserWindow.getAllWindows()[0] ?? null;
    const options: Electron.OpenDialogOptions = {
      title: 'Choose background image',
      properties: ['openFile'],
      filters: [{ name: 'Images', extensions: IMAGE_TYPES.map((t) => t.ext) }]
    };
    const picked = focused
      ? await dialog.showOpenDialog(focused, options)
      : await dialog.showOpenDialog(options);
    if (picked.canceled || picked.filePaths.length === 0) return { ok: true, data: null };
    return setWallpaperFromPath(picked.filePaths[0]);
  } catch (e) {
    return { ok: false, error: errMsg(e) };
  }
}

/**
 * Remove the active wallpaper: empty the setting, then delete the file. Runs
 * inside the mutation queue (same read→persist→delete shape as activate, and
 * the same orphan race if a clear interleaves a set).
 */
export function clearWallpaper(): Promise<Result<void>> {
  return enqueueWallpaperMutation(async () => {
    try {
      const previous = await currentFileName();
      const persisted = await setSettings({ wallpaperFileName: '' });
      if (!persisted.ok) return persisted;
      if (previous) await deleteManagedWallpaper(wallpaperDir(), previous);
      return { ok: true, data: undefined };
    } catch (e) {
      return { ok: false, error: errMsg(e) };
    }
  });
}

File: src-tauri/src/wallpaper.rs
//! wallpaper — port of src/main/wallpaper.ts (task #3, owner: worker-persist).
//!
//! To preserve: managed {userData}/wallpaper/ folder, wallpaper-<ms>.<ext>
//! names, raster-only png/jpg/jpeg/webp/gif/bmp/avif, 20MB cap, http(s) +
//! content-type-gated download (reqwest streaming), base64 `data:` URL serve,
//! name regex validation, write-new → flip-setting → delete-old order,
//! serialized mutations. pick() cancel resolves Ok(null).

use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;

use base64::Engine as _;
use tauri::Emitter;
use tauri::Manager;

use crate::contract::{Settings, WallpaperState};
use crate::result::NpResult;

/// Managed subfolder under userData holding the (single) active wallpaper.
const WALLPAPER_DIR_NAME: &str = "wallpaper";

/// Hard size cap for any wallpaper source (local stat / download stream).
const MAX_WALLPAPER_BYTES: u64 = 20 * 1024 * 1024; // 20 MB

/// Allowed raster image extensions.
const ALLOWED_EXTENSIONS: &[&str] = &["png", "jpg", "jpeg", "webp", "gif", "bmp", "avif"];

// ---------------------------------------------------------------------------
//  Pure helpers
// ---------------------------------------------------------------------------

/// MIME types mapped by canonical extension.
fn mime_for_extension(ext: &str) -> Option<&'static str> {
    match ext.to_lowercase().as_str() {
        "png" => Some("image/png"),
        "jpg" | "jpeg" => Some("image/jpeg"),
        "webp" => Some("image/webp"),
        "gif" => Some("image/gif"),
        "bmp" => Some("image/bmp"),
        "avif" => Some("image/avif"),
        _ => None,
    }
}

/// Map a Content-Type to the canonical extension, or None when not allowed.
fn extension_for_content_type(content_type: &str) -> Option<&'static str> {
    let mime = content_type.split(';').next()?.trim().to_lowercase();
    match mime.as_str() {
        "image/png" => Some("png"),
        "image/jpeg" => Some("jpg"),
        "image/webp" => Some("webp"),
        "image/gif" => Some("gif"),
        "image/bmp" => Some("bmp"),
        "image/avif" => Some("avif"),
        _ => None,
    }
}

/// Lower-cased extension (no dot) of a path/name, "" when absent.
fn image_extension_of(path_or_name: &str) -> String {
    std::path::Path::new(path_or_name)
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_lowercase()
}

/// True when ext (no dot, any case) is an allowed raster image extension.
fn is_allowed_image_extension(ext: &str) -> bool {
    ALLOWED_EXTENSIONS.contains(&ext.to_lowercase().as_str())
}

/// Compose a managed file name: `wallpaper-<ms>.<ext>`.
fn build_wallpaper_file_name(ext: &str, now_ms: u128) -> String {
    format!("wallpaper-{now_ms}.{}", ext.to_lowercase())
}

/// Validate a PERSISTED wallpaper file name. Must match `wallpaper-<digits>.<allowed_ext>`.
fn is_safe_wallpaper_file_name(name: &str) -> bool {
    let parts: Vec<&str> = name.splitn(2, '.').collect();
    if parts.len() != 2 {
        return false;
    }
    let stem = parts[0];
    let ext = parts[1];

    if !stem.starts_with("wallpaper-") || stem.len() <= "wallpaper-".len() {
        return false;
    }
    let num_part = &stem["wallpaper-".len()..];
    if num_part.is_empty() || !num_part.chars().all(|c| c.is_ascii_digit()) {
        return false;
    }
    is_allowed_image_extension(ext)
}

// ---------------------------------------------------------------------------
//  Path helpers
// ---------------------------------------------------------------------------

fn app_data_dir(app: &tauri::AppHandle) -> PathBuf {
    if let Ok(ov) = std::env::var("NOTEPADS_E2E_USERDATA") {
        if !ov.is_empty() {
            return PathBuf::from(ov);
        }
    }
    app.path()
        .app_data_dir()
        .expect("app_data_dir should exist")
}

fn wallpaper_dir(app: &tauri::AppHandle) -> PathBuf {
    app_data_dir(app).join(WALLPAPER_DIR_NAME)
}

fn settings_file_path(app: &tauri::AppHandle) -> PathBuf {
    app_data_dir(app).join("Settings.json")
}

/// Read the current wallpaper file name from settings.
fn current_file_name(app: &tauri::AppHandle) -> String {
    crate::settings::settings_get_internal(app)
        .map(|s| s.wallpaper_file_name)
        .unwrap_or_default()
}

// ---------------------------------------------------------------------------
//  Mutation serializer
// ---------------------------------------------------------------------------

/// Mutex that serializes wallpaper mutations (set/clear) to prevent orphan
/// races. Only the read→persist→delete critical section is serialized;
/// slow downloads/copies happen outside.
static WALLPAPER_MUTEX: Mutex<()> = Mutex::new(());

/// Persist settings atomically and broadcast the change.
fn persist_settings(app: &tauri::AppHandle, settings: &Settings) -> Result<(), String> {
    let serialized = serde_json::to_string_pretty(settings)
        .map_err(|e| format!("Failed to serialize settings: {e}"))?;

    let target_path = settings_file_path(app);
    let pid = std::process::id();
    let tmp_path = target_path.with_file_name(format!("Settings.json.{pid}.tmp"));

    if let Some(parent) = tmp_path.parent() {
        fs::create_dir_all(parent).map_err(|e| format!("Failed to create dir: {e}"))?;
    }

    fs::write(&tmp_path, &serialized).map_err(|e| format!("Failed to write tmp: {e}"))?;
    fs::rename(&tmp_path, &target_path).map_err(|e| {
        let _ = fs::remove_file(&tmp_path);
        format!("Failed to rename: {e}")
    })?;

    if let Err(e) = app.emit("notepads:evt:settings:changed", settings) {
        log::warn!("Failed to broadcast settings changed: {e}");
    }

    Ok(())
}

// ---------------------------------------------------------------------------
//  Managed-file operations
// ---------------------------------------------------------------------------

fn write_managed_file(
    dir: &std::path::Path,
    file_name: &str,
    bytes: &[u8],
) -> Result<PathBuf, String> {
    fs::create_dir_all(dir).map_err(|e| format!("Failed to create dir: {e}"))?;
    let target = dir.join(file_name);
    fs::write(&target, bytes).map_err(|e| format!("Failed to write file: {e}"))?;
    Ok(target)
}

fn delete_managed_file(dir: &std::path::Path, file_name: &str) {
    if file_name.is_empty() || !is_safe_wallpaper_file_name(file_name) {
        return;
    }
    let _ = fs::remove_file(dir.join(file_name));
}

/// Read wallpaper state: file name + base64 data URL.
fn read_wallpaper_state(app: &tauri::AppHandle) -> Result<WallpaperState, String> {
    let file_name = current_file_name(app);
    if file_name.is_empty() || !is_safe_wallpaper_file_name(&file_name) {
        return Ok(WallpaperState {
            file_name: String::new(),
            data_url: None,
        });
    }

    let ext = image_extension_of(&file_name);
    let mime = match mime_for_extension(&ext) {
        Some(m) => m,
        None => {
            return Ok(WallpaperState {
                file_name: String::new(),
                data_url: None,
            })
        }
    };

    let path = wallpaper_dir(app).join(&file_name);
    let bytes = match fs::read(&path) {
        Ok(b) => b,
        Err(_) => {
            return Ok(WallpaperState {
                file_name: String::new(),
                data_url: None,
            });
        }
    };

    let b64 = base64::engine::general_purpose::STANDARD.encode(&bytes);
    Ok(WallpaperState {
        file_name,
        data_url: Some(format!("data:{mime};base64,{b64}")),
    })
}

/// Activate a new wallpaper: persist in settings, then delete previous file.
fn activate(app: &tauri::AppHandle, file_name: &str) -> Result<WallpaperState, String> {
    let previous = {
        let _guard = WALLPAPER_MUTEX.lock().unwrap_or_else(|p| p.into_inner());
        let prev = current_file_name(app);

        let mut settings = crate::settings::settings_get_internal(app)?;
        settings.wallpaper_file_name = file_name.to_string();
        persist_settings(app, &settings)?;

        prev
    };

    if !previous.is_empty() && previous != file_name {
        delete_managed_file(&wallpaper_dir(app), &previous);
    }

    read_wallpaper_state(app)
}

// ---------------------------------------------------------------------------
//  Inner impls (return Result<T, String> for ? operator use)
// ---------------------------------------------------------------------------

fn set_from_path_impl(app: &tauri::AppHandle, path: &str) -> Result<WallpaperState, String> {
    let ext = image_extension_of(path);
    if !is_allowed_image_extension(&ext) {
        let label = if ext.is_empty() { "(none)" } else { &ext };
        return Err(format!("Unsupported image type: .{label}"));
    }

    let meta = fs::metadata(path).map_err(|e| format!("Failed to stat file: {e}"))?;
    if !meta.is_file() {
        return Err("Not a file".into());
    }
    if meta.len() > MAX_WALLPAPER_BYTES {
        let mb = MAX_WALLPAPER_BYTES / (1024 * 1024);
        return Err(format!("Image exceeds the {mb}MB limit"));
    }

    let now_ms = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis();
    let file_name = build_wallpaper_file_name(&ext, now_ms);
    let dir = wallpaper_dir(app);

    let bytes = fs::read(path).map_err(|e| format!("Failed to read image: {e}"))?;
    write_managed_file(&dir, &file_name, &bytes)?;

    activate(app, &file_name)
}

async fn set_from_url_impl(app: &tauri::AppHandle, url: &str) -> Result<WallpaperState, String> {
    let parsed = url::Url::parse(url).map_err(|_| "Invalid URL".to_string())?;
    if parsed.scheme() != "http" && parsed.scheme() != "https" {
        return Err("Only http(s) URLs are supported".into());
    }

    let client = reqwest::Client::builder()
        .build()
        .map_err(|e| format!("Failed to create HTTP client: {e}"))?;

    let response = client
        .get(url.to_string())
        .send()
        .await
        .map_err(|e| format!("Download failed: {e}"))?;

    let status = response.status();
    if !status.is_success() {
        return Err(format!("Download failed (HTTP {})", status.as_u16()));
    }

    let content_type = response
        .headers()
        .get("content-type")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");
    let ext = extension_for_content_type(content_type)
        .ok_or_else(|| "URL did not return a supported image type".to_string())?;

    if let Some(len) = response.content_length() {
        if len > MAX_WALLPAPER_BYTES {
            let mb = MAX_WALLPAPER_BYTES / (1024 * 1024);
            return Err(format!("Image exceeds the {mb}MB limit"));
        }
    }

    let bytes = response
        .bytes()
        .await
        .map_err(|e| format!("Download stream error: {e}"))?;

    if bytes.is_empty() {
        return Err("Empty response body".into());
    }
    if bytes.len() as u64 > MAX_WALLPAPER_BYTES {
        let mb = MAX_WALLPAPER_BYTES / (1024 * 1024);
        return Err(format!("Image exceeds the {mb}MB limit"));
    }

    let now_ms = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis();
    let file_name = build_wallpaper_file_name(ext, now_ms);
    let dir = wallpaper_dir(app);
    write_managed_file(&dir, &file_name, &bytes)?;

    activate(app, &file_name)
}

// ---------------------------------------------------------------------------
//  Tauri commands
// ---------------------------------------------------------------------------

#[tauri::command]
pub async fn wallpaper_get(app: tauri::AppHandle) -> NpResult<WallpaperState> {
    NpResult::from_result(read_wallpaper_state(&app))
}

#[tauri::command]
pub async fn wallpaper_set_from_path(
    app: tauri::AppHandle,
    path: String,
) -> NpResult<WallpaperState> {
    NpResult::from_result(set_from_path_impl(&app, &path))
}

#[tauri::command]
pub async fn wallpaper_set_from_url(
    app: tauri::AppHandle,
    url: String,
) -> NpResult<WallpaperState> {
    NpResult::from_result(set_from_url_impl(&app, &url).await)
}

#[tauri::command]
pub async fn wallpaper_pick(
    app: tauri::AppHandle,
    window: tauri::WebviewWindow,
) -> NpResult<Option<WallpaperState>> {
    use tauri_plugin_dialog::DialogExt;

    let result = app
        .dialog()
        .file()
        .set_parent(&window)
        .add_filter("Images", ALLOWED_EXTENSIONS)
        .blocking_pick_file();

    match result {
        Some(file_path) => {
            let path = file_path.to_string();
            match wallpaper_set_from_path(app, path).await {
                NpResult::Ok(state) => NpResult::Ok(Some(state)),
                NpResult::Err(e) => NpResult::Err(e),
            }
        }
        None => NpResult::Ok(None),
    }
}

#[tauri::command]
pub async fn wallpaper_clear(app: tauri::AppHandle) -> NpResult<()> {
    NpResult::from_result(clear_wallpaper_internal(&app))
}

/// Internal clear for settings_reset.rs.
pub fn clear_wallpaper_internal(app: &tauri::AppHandle) -> Result<(), String> {
    let previous = {
        let _guard = WALLPAPER_MUTEX.lock().unwrap_or_else(|p| p.into_inner());
        let prev = current_file_name(app);

        let mut settings = crate::settings::settings_get_internal(app)?;
        settings.wallpaper_file_name = String::new();
        persist_settings(app, &settings)?;

        prev
    };

    if !previous.is_empty() {
        delete_managed_file(&wallpaper_dir(app), &previous);
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_image_extension_of() {
        assert_eq!(image_extension_of("C:/pics/photo.PNG"), "png");
        assert_eq!(image_extension_of("/home/u/a.jpeg"), "jpeg");
        assert_eq!(image_extension_of("noext"), "");
    }

    #[test]
    fn test_is_allowed_image_extension() {
        for ext in ["png", "jpg", "jpeg", "webp", "gif", "bmp", "avif", "PNG"] {
            assert!(is_allowed_image_extension(ext), "should allow {ext}");
        }
        for ext in ["svg", "exe", "html", "js", ""] {
            assert!(!is_allowed_image_extension(ext), "should reject {ext}");
        }
    }

    #[test]
    fn test_mime_for_extension() {
        assert_eq!(mime_for_extension("png"), Some("image/png"));
        assert_eq!(mime_for_extension("jpg"), Some("image/jpeg"));
        assert_eq!(mime_for_extension("svg"), None);
    }

    #[test]
    fn test_extension_for_content_type() {
        assert_eq!(extension_for_content_type("image/png"), Some("png"));
        assert_eq!(extension_for_content_type("image/jpeg"), Some("jpg"));
        assert_eq!(extension_for_content_type("IMAGE/WebP"), Some("webp"));
        assert_eq!(
            extension_for_content_type("image/png; charset=binary"),
            Some("png")
        );
        assert_eq!(extension_for_content_type("text/html"), None);
        assert_eq!(extension_for_content_type("image/svg+xml"), None);
        assert_eq!(extension_for_content_type(""), None);
    }

    #[test]
    fn test_build_wallpaper_file_name() {
        let name = build_wallpaper_file_name("PNG", 1234567890);
        assert_eq!(name, "wallpaper-1234567890.png");
        assert!(is_safe_wallpaper_file_name(&name));
    }

    #[test]
    fn test_successive_names_differ() {
        assert_ne!(
            build_wallpaper_file_name("png", 1),
            build_wallpaper_file_name("png", 2)
        );
    }

    #[test]
    fn test_is_safe_wallpaper_file_name_rejects_bad_names() {
        assert!(!is_safe_wallpaper_file_name("../Settings.json"));
        assert!(!is_safe_wallpaper_file_name("wallpaper-1/..\\x.png"));
        assert!(!is_safe_wallpaper_file_name("wallpaper-1.svg"));
        assert!(!is_safe_wallpaper_file_name("wallpaper-1.exe"));
        assert!(!is_safe_wallpaper_file_name("other-1.png"));
        assert!(!is_safe_wallpaper_file_name(""));
    }

    #[test]
    fn test_max_wallpaper_bytes() {
        assert_eq!(MAX_WALLPAPER_BYTES, 20 * 1024 * 1024);
    }
}
